src/RestControllers/FHIR/FhirExportRestController.php

```diff
@@ -0,0 +1,574 @@
+
+            if (!empty($output)) {
+                $outputResult[] = $output;
+            } else {
+                $errorResult[] = $error;
+            }
+        }
+
+        $processedJob->setOutput(json_encode($outputResult));
+        $processedJob->setErrors(json_encode($errorResult));
+        $processedJob->setStatus(ExportJob::STATUS_COMPLETED);
+
+        return $this->fhirExportJobService->updateJob($processedJob);
+    }
+
+    private function createErrorResultForOutcomeOperation(ExportJob $job, FHIROperationOutcome $errorOutcome)
+    {
+        $jobUuidString = $job->getUuidString();
+        $resource = $errorOutcome->get_fhirElementName();
+        $fileName = $resource  . "-" . $jobUuidString . "-" . $errorOutcome->getId() . ".ndjson";
+        $data = json_encode($errorOutcome);
+        $document = $this->createExportJobFile($job, $fileName, $data);
+        return $this->getResultForResourceDocument($resource, $document);
+    }
+
+    private function createOutputResultForData(ExportJob $job, $resource, &$data)
+    {
+        $jobUuidString = $job->getUuidString();
+        $fileName = $resource . "-" . $jobUuidString . ".ndjson";
+        $document = $this->createExportJobFile($job, $fileName, $data);
+        return $this->getResultForResourceDocument($resource, $document);
+    }
+
+    private function getResultForResourceDocument($resource, \Document $document)
+    {
+        return [
+            'url' => $this->request->getApiBaseFullUrl() . '/fhir/Document/' . $document->get_id() . '/Binary'
+            ,"type" => $resource
+        ];
+    }
+
+    private function createExportJobFile(ExportJob $job, $fileName, &$data): \Document
+    {
+        $document = new \Document();
+        $folder = self::FHIR_DOCUMENT_FOLDER;
+        $categoryId = sqlQuery('Select `id` FROM categories WHERE name=?', [self::FHIR_DOCUMENT_CATEGORY]);
+        if ($categoryId === false) {
+            throw new ExportException("document category id does not exist in system");
+        }
+
+        $mimeType = "application/fhir+ndjson";
+        $higherLevelPath = "";
+        $pathDepth = 1;
+        $owner = 0;  // userID
+        $tmpFile = null;
+
+        $expirationDate = $job->getStartTime()
+            ->add(new \DateInterval(self::MAX_DOCUMENT_ACCESS_TIME))
+            ->format("Y-m-d H:i:s");
+
+        // set the foreign key so we can track documents connected to a specific export
+        $result = $document->createDocument(
+            $folder,
+            $categoryId,
+            $fileName,
+            $mimeType,
+            $data,
+            $higherLevelPath,
+            $pathDepth,
+            $owner,
+            $tmpFile,
+            $expirationDate
+        );
+        if (!empty($result)) {
+            throw new \RuntimeException("Failed to save document for job. Message: " . $result);
+        }
+        // not sure why we have to double save but the document class is broken in handling foreign ids
```

Comment:
note that when using this mechanism ($folder for patient id in createDocument class  (which is the correct way to do what you are doing), this mechanism will set the foreign id to 0 since it assumes it is not matched to a patient)

Reply:
I removed the comment since we handle foreign keys now a different way for documents.